### ISSUE TITLE
Missing default causes infinite recursion on repr

### ISSUE THREAD

**Issue by vidartf** [jupyter-widgets/ipywidgets#2835]
Using this code:

```python
from ipywidgets import Widget
from traitlets import Enum

class Foo(Widget):
     enum = Enum(['foo', 'bar']).tag(sync=True)

Foo()
```

You get an infinite recursion because of the missing default value of the `enum` trait:
- On init (`widget.open`) we try to evaluate the `enum` trait, traitlets detects the missing default, and raises an error: https://github.com/ipython/traitlets/blob/ba0ff90a2220ecf15062594428c3794d232c4cde/traitlets/traitlets.py#L533-L534
- The error message includes a `repr()` of the widget.
- Our custom `repr()` tries to enumerate all our traits:
  https://github.com/jupyter-widgets/ipywidgets/blob/ac2032d2f9cb930025979462aaccac4049721648/ipywidgets/widgets/widget.py#L562-L563
  into 
  https://github.com/jupyter-widgets/ipywidgets/blob/ac2032d2f9cb930025979462aaccac4049721648/ipywidgets/widgets/widget.py#L691

Proposed fix:
Introduce a guard in our `__repr__` method to prevent recursion. This is probably healthy for a variety of other cases as well.

**Comment by maartenbreddels** [comment]
Complex repr's have been a constant pain in vaex as well, also causes pytest to misbehave, since it calls repr as well. I think your idea sounds good, what should repr return in that case?

**Comment by vidartf** [comment]
> what should repr return in that case?

Probably class name and the model id if it has been set (typically not the case), if not, maybe just `id(self)` ?

**Comment by maartenbreddels** [comment]
Maybe with some indicator that we found a recursive call? Since this shouldn't normally happen right? Or would you say, it only happens on an exception, so it doesn't matter so much?

**Comment by vidartf** [comment]
ie:
```python
    def _gen_repr_from_keys(self, keys):
        class_name = self.__class__.__name__
        if self._repr_guard:
            if self._model_id:
                return '%s[model_id=%r]' % (class_name, self._model_id)
            return '%s<%s>' % (class_name, id(self))
        self._repr_guard = True
        try:
            signature = ', '.join(
                '%s=%r' % (key, getattr(self, key))
                for key in keys
            )
            return '%s(%s)' % (class_name, signature)
        finally:
            self._repr_guard = False
```

**Comment by vidartf** [comment]
Yeah, a "recursive" indicator could be useful, probably as a prefix `Recursive repr: ...`

**Comment by vidartf** [comment]
> also causes pytest to misbehave, since it calls repr as well

I assume this only happens for failing tests?

**Comment by maartenbreddels** [comment]
Indeed.
